### ISSUE TITLE
spawn nix-instantiate ENOENT

### ISSUE THREAD

**Issue by andrewbaxter** [nix-community/vscode-nix-ide#428]
Same as #404 (closed).

I just installed the plugin and immediately got `spawn nix-instantiate ENOENT` when I opened a `.nix` file.  I installed `nixd` per @inclyc 's link in #404 and the error message went away.  I don't have any settings set, like `"nix.enableLanguageServer": true,`.

The presence of an error message popup I believe unambiguously means that the plugin expected something to work and it failed unexpectedly, and notified the user because it might (again, unexpectedly) impact their usage.

I'm not sure I 100% understand the discussion in #404. Can someone confirm my summary?

So I think either

1. A language server is required, and the installation section of the `Nix IDE` should be updated to reflect this

2. A language server shouldn't be required, and there's some bug in the plugin that's not checking for its presence

If it's 1 (and FWIW I think it's reasonable to require an LSP, developer machines are powerful and it might simplify things) I'd be glad to make an MR.

I think a good use case is: A new person joins a company using Nix for images or whatever.  They've just set up their computer and come across a Nix file in some repo, and go to the extensions marketplace for an extension and find `Nix IDE`.  They've never used Nix before and don't have anything set up, so they read the extension readme to figure out what they need to do.  They are neither using a Nix system nor are they using Nix as a package manager on a foreign distro currently (maybe they will eventually, but they're relying on CI for now).

Since extensions are often people's entry points to development ecosystems, in this case I think it wouldn't be crazy to list _all_ requirements: Nix itself, LSP, plus the formatter?  Just linking to their installation instructions, not duplicating anything.

Edit: Saving a file shows `failed to run nixpkgs-fmt: spawn nixpkgs-fmt ENOENT` and a popup about waiting for Nix IDE that lasts indefinitely so it seems like `nixpkgs-fmt` is non-optional too.

**Comment by jnoortheen** [comment]
`nixpkgs-fmt` is used for auto-formatting the file (only when LSP is not enabled). If that is not available in PATH, then the error will be shown. So you may have configured for `format on save` option. Closing this as it is an expected behaviour.

**Comment by filiptrplan** [comment]
> `nixpkgs-fmt` is used for auto-formatting the file (only when LSP is not enabled). If that is not available in PATH, then the error will be shown. So you may have configured for `format on save` option. Closing this as it is an expected behaviour.

I think this should be reopened. I have it in my path, the binary works and I even set the `nix.formatterPath' to the exact absolute path of the binary and I'm getting the same error.

I have `nixpkgs-fmt` installed using Home Manager.

**Comment by OliverJAsh** [comment]
I'm running into the same issue when formatting files, although the error message is slightly different: `Failed to run nixfmt: spawn nixfmt ENOENT`.

FWIW, I do have `nixpkgs-fmt` in my `PATH`.

**Comment by andrewbaxter** [comment]
Yes I ran into it again on a new system just this week.  For me installing `nixpkgs-fmt` wasn't enough, I also needed `nixd`.

This in the readme

> Formatting the code should work if [nixfmt](https://github.com/NixOS/nixfmt) (or the archived [nixpkgs-fmt](https://github.com/nix-community/nixpkgs-fmt)) is installed and available on the $PATH.

is wrong, or at the very least is heavily dependent on unstated subtleties, or there's a bug.

**Comment by OliverJAsh** [comment]
Ah, I just installed `nixfmt` and uninstalled `nixpkgs-fmt`, and it fixed the issue for me.

**Comment by rverma-sf** [comment]
I am facing same issue when using alejendra
  "nix.formatterPath" = "alejandra";
  "nix.serverPath"= "nil";
I even tried with nixd as serverPath and installing the nixd but it didn't helped

**Comment by bitscoper** [comment]
> Ah, I just installed `nixfmt` and uninstalled `nixpkgs-fmt`, and it fixed the issue for me.

It worked; thanks, @OliverJAsh.
